### PR TITLE
[Merged by Bors] - feat: more on central simple algebras

### PR DESCRIPTION
**Pull request by Whysoserioushah**
Co-authored-by: Jujian Zhang <jujian.zhang1998@outlook.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Central/Basic.lean
Comment (kbuzzard) [inline Mathlib/Algebra/Central/Basic.lean:64]: This docstring is in some sense worse than no docstring, you can get more information by just reading the statement of the theorem.

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (kbuzzard) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:71]: All the previous results were in a section and used completely different variables. The point of sections is to control the scope of variables. Probably you should have let the section end here and started a new one if you had no interest in the variables already set up. Note also that all your lemmas have variables `{ιC ιB: Type*} (𝒞 : Basis ιC K C) (ℬ : Basis ιB K B)` so perhaps they should also be variables in your new section.

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (kbuzzard) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:76]: This statement looks quite weird. The assumption that B has a basis is not used in the statement and is not necessary in the proof; indeed all that one needs is that iC spans C as a K-vector space. Then one can write x=sum_i b_i (x) c_i, one writes each c_i in terms of the spanning set and tidies up. No hypotheses on B are necessary at all. It also feels like the most natural statement would be with a finsupp ιC →_0 B.

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (kbuzzard) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:116]: K is commutative so surely you can just deduce this from the previous lemma via [`Algebra.TensorProduct.comm`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/TensorProduct/Basic.html#Algebra.TensorProduct.comm) rather than going through the long proof again.

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (kbuzzard) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:155]: This is probably more nicely expressed as a map from finsupp being injective. In fact it seems to just be the statement that if ιC is K-linearly independent in C then it's B-linearly independent in B (x) C when B is free. It feels like this should be true more generally. What's the correct statement here?

File: Mathlib/Algebra/Central/Basic.lean
Comment (Whysoserioushah) [reply Mathlib/Algebra/Central/Basic.lean:64]: hmmm okay deleted

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (Whysoserioushah) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean:76]: Ah okay but I thought we need both basis for Basis.tensorproduct?

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (Whysoserioushah) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean:71]: good point I've adapted the original variable names now

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean]: In `TensorProduct.eq_repr_basis_right`, I used another proof avoiding `Basis.tensorProduct`; however in `TensorProduct.sum_tmul_basis_right_eq_zero`, I have to assume `M` is Free so tha tI can use `Basis.tensorProduct`

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean]: I assumed `B` is free, not sure what is the correct generality though

File: Mathlib/Algebra/Central/Basic.lean
Comment (jjaassoonn) [reply Mathlib/Algebra/Central/Basic.lean:69]: I think if this is in a new file, then the `large-import` part should be okay

Comment (jjaassoonn) [reply Mathlib/Algebra/Central/Basic.lean:69]: Can you prove the more general statement that `Matrix \iota \iota D` is central when `D` is central

File: Mathlib/LinearAlgebra/FreeModule/Basic.lean
Comment (jjaassoonn) [inline Mathlib/LinearAlgebra/FreeModule/Basic.lean:183]:
```suggestion
```

File: Mathlib/Algebra/Central/Basic.lean
Comment (jjaassoonn) [reply Mathlib/Algebra/Central/Basic.lean]: I have moved it

**Comment by jjaassoonn** [comment]
> ### PR summary [9a74e2169b](https://github.com/leanprover-community/mathlib4/pull/17789/commits/9a74e2169b95e5d9d001a2f6a9c115bd2ed3a45e)
> Import changes exceeding 2%
> 
> %	File
> +2.21%	`Mathlib.Algebra.Central.Basic`
> #### Import changes for modified files
> #### Declarations diff

Can we run this again, I think it shouldn't be `large-import` now

File: Mathlib/Algebra/Central/Matrix.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Central/Matrix.lean:6]: We don't usually put the space after the copyright header
```suggestion
```

File: Mathlib/Algebra/Central/Matrix.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Central/Matrix.lean:14]:
```suggestion
/-!
# The matrix algebra is a central algebra
-/
```

File: Mathlib/Algebra/Central/Matrix.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Central/Matrix.lean:25]: `where` notation lets you introduce variables before the `:=`
```suggestion
  out m h := by
```

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:73]:
```suggestion
/-- Elements in `M ⊗ N` can be represented by sum of elements in `M` tensor elements of basis of
`N`. -/
```

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:95]: Same here

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:75]:
```suggestion
    ∃ b : κ →₀ M, b.sum (fun i m => m ⊗ₜ 𝒞 i) = x := by
```
is possibly clearer

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (Whysoserioushah) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean]: thanks, applied

File: Mathlib/Algebra/Central/Matrix.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Central/Matrix.lean:15]: Why have this section given that there is only result in it?

File: Mathlib/LinearAlgebra/FreeModule/Basic.lean
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/FreeModule/Basic.lean:210]:
```suggestion
          congr! with j h
```
same everywhere else

**Review by YaelDillies — APPROVED** [review]
Thanks for the PR! It looks good to me modulo a few comments, but an algebraist will need to approve as well

maintainer merge

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:81]:
```suggestion
    use (𝒞.repr n).mapRange (fun (r : R) ↦ r • m) (by simp)
```
The `.|>` trick is nice but in my opinion makes things harder to read.

Similarly elsewhere.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

**Comment by jjaassoonn** [comment]
Let's not merge before the following comment is resolved
> I don't think you need any hypotheses on B at all. If C is free then B tensor C is just B tensor (finsupp I R) which is finsupp I B and now injectivity is immediate. This probably shortens the proof too.

File: Mathlib/Algebra/Central/Matrix.lean
Comment (jjaassoonn) [reply Mathlib/Algebra/Central/Matrix.lean]: indeed

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean]: Changed thanks

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean]: I am now pretty satisfied with the current approach, can you look at `LinearAlgebra/TensorProduct/Basis.lean` line 71-137 again

**Comment by jjaassoonn** [comment]
> Let's not merge before the following comment is resolved
> 
> > I don't think you need any hypotheses on B at all. If C is free then B tensor C is just B tensor (finsupp I R) which is finsupp I B and now injectivity is immediate. This probably shortens the proof too.

I am now pretty satisfied with the current approach, can you look at LinearAlgebra/TensorProduct/Basis.lean line 71-137 again @riccardobrasca

File: Mathlib/LinearAlgebra/TensorProduct/Basis.lean
Comment (kbuzzard) [inline Mathlib/LinearAlgebra/TensorProduct/Basis.lean:77]: I would have gone for
```lean
def TensorProduct.equivFinsuppOfBasisRight : M ⊗[R] N ≃ₗ[R] κ →₀ M := by
  classical -- for DecidableEq κ
  exact LinearEquiv.lTensor M 𝒞.repr ≪≫ₗ TensorProduct.finsuppScalarRight R M κ
```

instead of your 25 lines. But this breaks the next proof, annoyingly. A fix is
```lean
lemma TensorProduct.equivFinsuppOfBasisRight_symm_apply (b : κ →₀ M) :
    (TensorProduct.equivFinsuppOfBasisRight 𝒞).symm b = b.sum fun i m ↦ m ⊗ₜ 𝒞 i := by
  classical
  induction b using Finsupp.induction_linear
  · simp
  · simp_all [Finsupp.sum_add_index, add_tmul]
  · simp [equivFinsuppOfBasisRight]
```

I don't have a good feeling about why `rfl` fails. Ultimately do you think that this approach is a win?

Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct/Basis.lean:78]: as long as `TensorProduct.equivFinsuppOfBasisRight_symm_apply` works, all the proofs are very easy to fix. So I have adapted to your definition

instead of `classical -- for DecidableEq κ`, I just assumed `DecidableEq \kappa`

**Comment by riccardobrasca** [comment]
Is this ready to be merged?

**Comment by jjaassoonn** [comment]
I think so now. Thanks

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**
